### PR TITLE
feat: implement threshold for resetting password

### LINKED ISSUE #6033: Limit the number of emails an account can sent out in a given timeframe
One account yesterday tried a password reset requesting a reset 1000 times.

Please implement a threshhold for account resets and propose a threshhold for other emails in a certain time frame. Open relevant issues for this.

A threshhold for an account reset email should be 3 times per hour. After that the user should receive a message: "You have reached the threshhold for this action. Please try again in one hour."

### PR DESCRIPTION
**Pull request by shreyanshdwivedi**
Fixes #6033 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:
A threshold for an account reset email should be 3 times per hour. After that, the user receives a message: "You have reached the threshold for this action. Please try again in one hour."

### CODE REVIEW

**Comment by shreyanshdwivedi** [comment]
@uds5501 @mrsaicharan1 @kushthedude I'm not getting the time in the same format as it is getting stored in the db. For example, the time getting stored in db is - `2019-06-12 13:36:29.233271+05:30` and one I'm getting is `2019-06-12 13:36:29.227708+00:00`. Can you guys help me out to find what am I missing here?

**Comment by kushthedude** [comment]
@shreyanshdwivedi We are not using pytz for time now , As recent change made by @prateekj117 we are now using `SQLalchemy` inbuilt time stamp function, Check there if you find a way to retireve time from db

File: app/api/auth.py [hunk 1]
```diff
@@ -222,7 +226,14 @@ def reset_password_post():
         if user.was_registered_with_order:
             send_email_with_action(user, PASSWORD_RESET_AND_VERIFY, app_name=get_settings()['app_name'], link=link)
         else:
-            send_email_with_action(user, PASSWORD_RESET, app_name=get_settings()['app_name'], link=link)
+            previous_hour = datetime.now(pytz.utc) - timedelta(hours=1)
```
Comment (prateekj117) [inline app/api/auth.py]: @shreyanshdwivedi Try datetime.utcnow

Comment (shreyanshdwivedi) [reply app/api/auth.py]: @prateekj117 when I used `datetime.utcnow()`, I'm getting datetime without timezone - `2019-06-12 12:54:13.035101` - neither `00:00` nor `05:30`

Comment (prateekj117) [reply app/api/auth.py]: The now() call returns the correct time for local location, while the utcnow() call returns the time in the UTC time zone.

Comment (shreyanshdwivedi) [reply app/api/auth.py]: The time returned by both are absolutely correct. I just want to include my current timezone

**Comment by shreyanshdwivedi** [comment]
@niranjan94 so should I use flask_limiter here rather than this approach for `password-reset`?

**Comment by niranjan94** [comment]
> so should I use flask_limiter here rather than this approach for `password-reset`?

Yep. Correct. And the frontend can show an appropriate message to the user if it gets a `429` response code.

**Comment by shreyanshdwivedi** [comment]
@niranjan94 ok cool. But one thing, the issue mentioned -
>A threshhold for an account reset email should be 3 times per hour. After that the user should receive a message: "You have reached the threshhold for this action. Please try again in one hour."

But I don't think this type of message can be achieved by flask_limiter.

**Comment by niranjan94** [comment]
> But I don't think this type of message can be achieved by flask_limiter.

@shreyanshdwivedi 

https://github.com/fossasia/open-event-server/pull/6035#issuecomment-501293990

File: app/api/auth.py (hunk 1 above)
Comment (prateekj117) [reply app/api/auth.py]: Try datetime.utcnow().isoformat()

File: app/api/auth.py [hunk 2]
```diff
@@ -207,6 +208,7 @@ def resend_verification_email():
 
 
 @auth_routes.route('/reset-password', methods=['POST'])
+@limiter.limit("10/s", exempt_when=lambda: current_user.is_admin, error_message='Limit for this action exceeded')
```
Comment (iamareebjamal) [inline app/api/auth.py]: Use 3/hour

File: app/api/auth.py [hunk 3]
```diff
@@ -29,10 +30,10 @@
 from app.models.user import User
 from app.api.helpers.storage import UPLOAD_PATHS
 
-
 authorised_blueprint = Blueprint('authorised_blueprint', __name__, url_prefix='/')
 ticket_blueprint = Blueprint('ticket_blueprint', __name__, url_prefix='/v1')
 auth_routes = Blueprint('auth', __name__, url_prefix='/v1/auth')
+limiter = Limiter()
```
Comment (iamareebjamal) [inline app/api/auth.py]: Initialize in root __init__.py so all nested modules can use

**Comment by shreyanshdwivedi** [comment]
@iamareebjamal @niranjan94 I've updated the `key_func` and finalized the PR. I've checked it on local and it works for me. The error message produced -
```
429 Too Many Requests
Too Many Requests
Limit for this action exceeded
```
Please review

File: app/api/auth.py [hunk 4]
```diff
@@ -207,6 +207,9 @@ def resend_verification_email():
 
 
 @auth_routes.route('/reset-password', methods=['POST'])
+@limiter.limit(
+    '3/hour', key_func=lambda: request.json['data']['email'], error_message='Limit for this action exceeded'
```
Comment (iamareebjamal) [inline app/api/auth.py]: jamal.areeb@gmail.com
jamalareeb@gmail.com
ja.mal.are.eb@gmail.com

Are all same email addresses and will be allowed by this method

Comment (shreyanshdwivedi) [reply app/api/auth.py]: @iamareebjamal I just tested the above three email addresses.
Each one is distinguished perfectly and failed when the email if entered fourth time. I jumbled the order and tested, the reset request for one email is not affecting other in any way

Comment (iamareebjamal) [reply app/api/auth.py]: It should

Comment (iamareebjamal) [reply app/api/auth.py]: The point is that I can keep adding dots and sending emails at a rate of more than 3/hour

I can loop through a list of emails and send password reset request on them. Essentially brute forcing

Comment (shreyanshdwivedi) [reply app/api/auth.py]: But there is a check which raises error if user is not registered and so no mail is sent out. If a user loops by adding dots, it doesn't mean the email is registered and so error will be raised. I hope it makes sense to you.

Comment (iamareebjamal) [reply app/api/auth.py]: A hacker doesn't care if the error is raised. He'll just continue looping through emails, and our resources will get exhausted. And users will still get spammed.

Also, raising error if user is not registered is a pretty big security issue, it should be fixed ASAP - #6069

Comment (shreyanshdwivedi) [reply app/api/auth.py]: @iamareebjamal from what I understood, I should create a function which treats similar emails as same email. Right?
Also can you please guide me which pattern should I consider ?

Comment (iamareebjamal) [reply app/api/auth.py]: Instead, you should add one more limiter which prevents same IP to access the unprotected resource at a lower rate, like '1/minute'

Comment (shreyanshdwivedi) [reply app/api/auth.py]: @iamareebjamal I've updated the PR. Please review

**Pull request merged**